spylib/token.py

```diff
@@ -0,0 +1,428 @@
+from abc import ABC, abstractmethod
+from asyncio import sleep
+from datetime import datetime, timedelta
+from time import monotonic
+from typing import Any, Dict, List, Optional
+import json
+from loguru import logger
+from pydantic import BaseModel, root_validator, MissingError
+from pydantic.dataclasses import dataclass
+from pydantic.main import validate_model
+from shortuuid.main import ShortUUID  # type: ignore
+
+from math import ceil, floor
+from httpx import AsyncClient, Response
+from tenacity import retry
+from tenacity.retry import retry_if_exception, retry_if_exception_type
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_random
+
+from spylib.constants import (
+    MAX_COST_EXCEEDED_ERROR_CODE,
+    OPERATION_NAME_REQUIRED_ERROR_MESSAGE,
+    THROTTLED_ERROR_CODE,
+    WRONG_OPERATION_NAME_ERROR_MESSAGE,
+)
+
+from spylib.exceptions import (
+    ShopifyCallInvalidError,
+    ShopifyError,
+    ShopifyExceedingMaxCostError,
+    ShopifyThrottledError,
+    not_our_fault,
+)
+
+
+class AssociatedUser(BaseModel):
+    id: int
+    first_name: str
+    last_name: str
+    email: str
+    email_verified: bool
+    account_owner: bool
+    locale: str
+    collaborator: bool
+
+
+class OfflineTokenResponse(BaseModel):
+    access_token: str
+    scope: str
+
+
+class OnlineTokenResponse(BaseModel):
+    access_token: str
+    scope: str
+    expires_in: int
+    associated_user_scope: str
+    associated_user: AssociatedUser
+
+
+class Token(ABC, BaseModel):
+    """
+    Abstract class for token objects. This should never be extended, as you
+    should either be using the OfflineToken or the OnlineToken.
+    """
+
+    store_name: str
+    scope: List[str]
+    access_token: Optional[str]
+    access_token_invalid: bool
+    client: AsyncClient
+    oauth_url: str
+    api_url: str
+    api_version: str = '2021-04'
+    rest_bucket_max: int = 80
+    rest_leak_rate: int = 4
+    graphql_bucket_max: int = 1000
+    graphql_leak_rate: int = 50
```

Comment:
`ClassVar` here too. No point in the user setting these values. We need defaults but at the first API call the response header will give us the proper values. Much better way to do it.

Reply:
I can change this, but will there ever be a scenario where we have an application that needs to support both regular and Shopify plus? Because making it a class variable will remove the ability to support that scenario.